# Bear_BBR Signal - Data Prep

## Import Libs

In [ ]:
import pandas as pd 
import numpy as np
import os
from pandas import DataFrame, Series
import plotly.graph_objects as go

## Read CSV price data and get DataFrame

In [479]:
PATH = os.getcwd()
FILE = "../output/FE_GBPUSD_15mins_1yr_End_20260311.csv"
df = pd.read_csv(FILE)
df["Date"] = pd.DatetimeIndex(df["Date"], tz="US/Eastern")
df.set_index("Date", inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 24580 entries, 2025-03-11 17:15:00-04:00 to 2026-03-11 16:45:00-04:00
Data columns (total 46 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Open           24580 non-null  float64
 1   High           24580 non-null  float64
 2   Low            24580 non-null  float64
 3   Close          24580 non-null  float64
 4   Idx            24580 non-null  int64  
 5   Body           24580 non-null  float64
 6   Range          24580 non-null  float64
 7   UWick          24580 non-null  float64
 8   LWick          24580 non-null  float64
 9   Close_%High    24580 non-null  float64
 10  Open_%High     24580 non-null  float64
 11  Iday_Idx       24580 non-null  int64  
 12  Iday_High      24580 non-null  float64
 13  Iday_Low       24580 non-null  float64
 14  Iday_Range     24580 non-null  float64
 15  Close_%DHigh   24580 non-null  float64
 16  Open_%DHigh    24580 non-null  float64
 17  Yda

## Add extra features

In [461]:
def get_yra_std(row: Series, df: DataFrame, n: int):
    std = df["YRA_Diff"].at_time("17:15").iloc[row["Day_Idx"]-n+1:row["Day_Idx"]+1].std()
    return std

pd.options.display.max_rows = 100
cols = ['Idx','Day_Idx', 'Open', 'High', 'Low', 'Close', 'Body', 'Range', 'ATR',
        'Close_%High', 'Open_%High', 'Iday_High', 'Iday_Low', 
        'Iday_Range', 'ADR', 'Close_%DHigh', 'Open_%DHigh', 
        'Yday_High', 'Yday_Low', 'Yday_Range', 'YRA_Diff', 'YRA_STD', 'SMA4', 'SMA16', 'SMA32',
       'BB_Upper_16_2', 'BB_Lower_16_2', 'RSI', 'Sig_High', 'Sig_Low',
       'SMA_Trend', 'SMA16_Slope', 'SMA32_Slope', 'Bear_BBR', 'Shooting_Star',
       'Bear_Engulf', 'Dark_Cloud', 'IHR', 'S_R', 'BBU_BO', 
       'ILR', 'Bull_BBR', 'Hammer', 'Bull_Engulf', 'BBL_BO']
pd.options.display.max_columns = len(cols)
df["Yday_Range"] = df.eval("Yday_High - Yday_Low")
df["YRA_Diff"] = df.apply(lambda x: x["Yday_Range"] - x["ADR"], axis=1)
df["YRA_STD"] = df.apply(get_yra_std, axis=1, args=[df, 30])



In [462]:
bbu_reversal = df[cols]
bbu_reversal.query("Bear_BBR == True")

,Idx,Day_Idx,Open,High,Low,Close,Body,Range,ATR,Close_%High,Open_%High,Iday_High,Iday_Low,Iday_Range,ADR,Close_%DHigh,Open_%DHigh,Yday_High,Yday_Low,Yday_Range,YRA_Diff,YRA_STD,SMA4,SMA16,SMA32,BB_Upper_16_2,BB_Lower_16_2,RSI,Sig_High,Sig_Low,SMA_Trend,SMA16_Slope,SMA32_Slope,Bear_BBR,Shooting_Star,Bear_Engulf,Dark_Cloud,IHR,S_R,BBU_BO,ILR,Bull_BBR,Hammer,Bull_Engulf,BBL_BO
Date,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
2025-03-12 05:00:00-04:00,47,0,1.294210,1.294530,1.293750,1.293895,0.000315,0.000780,0.000939,0.814103,0.410256,1.295685,1.291365,0.004320,NaN,0.414352,0.341435,NaN,NaN,NaN,NaN,NaN,1.293565,1.292857,1.292971,1.294028,1.291687,57.242400,1.295685,1.291365,1,35.509439,6.123933,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-12 07:30:00-04:00,57,0,1.295655,1.296145,1.295100,1.295430,0.000225,0.001045,0.000943,0.684211,0.468900,1.296145,1.291365,0.004780,NaN,0.149582,0.102510,NaN,NaN,NaN,NaN,NaN,1.294770,1.293912,1.293191,1.295513,1.292310,63.957243,1.295685,1.291365,2,50.534430,28.026001,True,NaN,NaN,NaN,True,1.0,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-12 12:15:00-04:00,76,0,1.298070,1.298265,1.297430,1.297660,0.000410,0.000835,0.001290,0.724551,0.233533,1.299000,1.291365,0.007635,NaN,0.175508,0.121807,NaN,NaN,NaN,NaN,NaN,1.297094,1.295018,1.294655,1.297964,1.292071,65.991526,1.299000,1.291365,2,48.524029,46.140500,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-12 22:15:00-04:00,115,1,1.297265,1.297350,1.296885,1.297010,0.000255,0.000465,0.000436,0.731183,0.182796,1.297350,1.296075,0.001275,NaN,0.266667,0.066667,1.299000,1.291365,0.007635,NaN,NaN,1.296846,1.296547,1.296601,1.297143,1.295951,57.141452,NaN,NaN,1,13.473158,1.909152,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-13 04:30:00-04:00,140,1,1.296485,1.296485,1.295535,1.295910,0.000575,0.000950,0.000938,0.605263,0.000000,1.297350,1.294155,0.003195,NaN,0.450704,0.270736,1.299000,1.291365,0.007635,NaN,NaN,1.295755,1.295502,1.295990,1.296433,1.294571,50.937783,1.297350,1.296075,0,4.051684,-0.029842,True,NaN,NaN,NaN,NaN,4.0,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-03-06 16:15:00-05:00,24292,252,1.340140,1.340270,1.339910,1.339925,0.000215,0.000360,0.001030,0.958333,0.361111,1.340980,1.331155,0.009825,0.010079,0.107379,0.085496,1.338710,1.329750,0.008960,-0.001119,0.003966,1.339426,1.339091,1.337920,1.340132,1.338050,58.904509,1.340980,1.331155,2,19.978314,46.297646,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2026-03-09 02:00:00-04:00,24330,253,1.333365,1.334075,1.332485,1.332730,0.000635,0.001590,0.001371,0.845912,0.446541,1.335175,1.328310,0.006865,0.009884,0.356154,0.263656,1.341630,1.331155,0.010475,0.000591,0.003713,1.332796,1.330734,1.330619,1.333298,1.328170,55.186083,1.335175,1.328310,2,59.302734,0.835504,True,NaN,NaN,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2026-03-09 09:30:00-04:00,24360,253,1.337745,1.338270,1.336690,1.337575,0.000170,0.001580,0.001383,0.439873,0.332278,1.338270,1.328310,0.009960,0.009884,0.069779,0.052711,1.341630,1.331155,0.010475,0.000591,0.003713,1.336907,1.335188,1.333955,1.337667,1.332710,66.050386,1.336365,1.328310,2,56.383248,46.819677,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# Bear_BBR Signal Analysis

## Bearish BBR Metrics Function

Return a new dataframe with all occrrences of Bear_BBR signal and get additional metrics such as:
- Entry Price
- Max Gain / Loss
- Gain/Loss to:
    - SMA16,
    - SMA32,
    - BB_Lower,
    - SigLow,
    - First_Bullish_Signal (Bull_0)
    - Bull_Engulf
    - Intraday_Low_Reversal (ILR)
- Whether S/R was tested or broken
- Whether other bearish pattern occurred at the same time

In [527]:

def bearish_bbr_metrics(df: DataFrame, bear_bbr_signal_idx: list):
    """Return a new dataframe with signal metrics"""

    signal = []
    for signal_idx in bear_bbr_signal_idx:
        row = df.iloc[signal_idx]
        s = {}
        START = row.name
        TD = pd.Timedelta(minutes=15)
        s["Date"] = START
        s["Entry_Price"] = row["Close"] # signal price
        
        day = row["Day_Idx"]
        window = df.query(f"Day_Idx == {day}") # only get df data from the same FX day as signal
        before_signal_window = window[:START-TD] # get df data from start of day until just before signal
        signal_window = window[START:] # get df from signal until end of day (EOD)
        
        Max_High = signal_window[START+TD:]["High"].max() # Max high from signal until EOD
        Min_Low = signal_window[START+TD:]["Low"].min() # Min low from signal until EOD
        s["D_Max_Up"] = s["Entry_Price"] - Max_High # Max price gain from signal until EOD
        s["D_Max_Down"] = s["Entry_Price"] - Min_Low # Max price loss from signal until EOD
        D_Close = signal_window.iloc[-1]["Close"] # Closing price at EOD
        s["D_Gain"] = s["Entry_Price"] - D_Close # Gain at EOD from signal price


        if signal_window.query(f"Close > {row["High"]}").empty is False:
            # If price closed above signal price within the day, 
            # get the dataframe between start and failure time
            EXIT = signal_window[START:].query(f"Close > {row["High"]}").iloc[0].name
            entry_window = signal_window[START:EXIT]
        else:
            # If price didn't close above signal price within the day
            # get the dataframe from signal time until EOD
            entry_window = signal_window

        # Did the price go higher than the signal candle high
        if entry_window[START+TD:].query(f"High >= {row["High"]}").empty is False:
            s["Stop_Pips"] = s["Entry_Price"] - row["High"]
        else:
            s["Stop_Pips"] = 0
        
        s["Max_Up"] = s["Entry_Price"] - entry_window[START+TD:]["Close"].max()
        s["Max_Down"] = s["Entry_Price"] - entry_window[START+TD:]["Low"].min()
        

        # If the price fell to the SMA16 after the signal
        # calcuate the gain
        below_sma16 = entry_window.query("Low < SMA16")
        if below_sma16.empty is False:
            s["To_SMA16"] = s["Entry_Price"] - below_sma16.iloc[0]["SMA16"] 
        else: 
            s["To_SMA16"] = 0

        # If the price fell to the SMA32 after the signal
        # calcuate the gain
        below_sma32 = entry_window.query("Low < SMA32")
        if below_sma32.empty is False:
            s["To_SMA32"] = s["Entry_Price"] - below_sma32.iloc[0]["SMA32"]
        else:
            s["To_SMA32"] = 0

        # If the price fell to the BB_Lower_16_2 after the signal
        # calcuate the gain
        below_bbl = entry_window.query("Low < BB_Lower_16_2")
        if below_bbl.empty is False:
            s["To_BBL_16_2"] = s["Entry_Price"] - below_bbl.iloc[0]["BB_Lower_16_2"]
        else:
            s["To_BBL_16_2"] = 0

        # If the price fell to the Sig_Low after the signal
        # calcuate the gain
        below_sig_low = entry_window.query("Low < Sig_Low")
        if below_sig_low.empty is False:
            s["To_Sig_Low"] = s["Entry_Price"] - below_sig_low.iloc[0]["Sig_Low"]
        else:
            s["To_Sig_Low"] = 0

        below_sr = entry_window.query("S_R > 3")["Sig_Low"].unique()
        for i in range(len(below_sr)):
            s[f"To_S_{i}"] = s["Entry_Price"] - below_sr[i]


        # Check if bullish signals occured during winddow
        NULL_Timestamp = pd.Timestamp(0, tz="US/Eastern")
        to_bull_bbr = entry_window.query("Bull_BBR == True")
        if to_bull_bbr.empty is False:
            s["To_Bull_BBR"] = s["Entry_Price"] - to_bull_bbr.iloc[0]["Close"]
            to_bull_bbr_time_stamp = to_bull_bbr.iloc[0].name
        else:
            s["To_Bull_BBR"] = 0
            to_bull_bbr_time_stamp = NULL_Timestamp

        to_bull_engulf = entry_window.query("Bull_Engulf == True")
        if to_bull_engulf.empty is False:
            s["To_Bull_Engulf"] = s["Entry_Price"] - to_bull_engulf.iloc[0]["Close"]
            to_bull_engulf_time_stamp = to_bull_engulf.iloc[0].name
        else:
            s["To_Bull_Engulf"] = 0
            to_bull_engulf_time_stamp = NULL_Timestamp

        to_ilr = entry_window.query("ILR == True")
        if to_ilr.empty is False:
            s["To_ILR"] = s["Entry_Price"] - to_ilr.iloc[0]["Close"]
            to_ilr_time_stamp = to_ilr.iloc[0].name
        else:
            s["To_ILR"] = 0
            to_ilr_time_stamp = NULL_Timestamp

        first_bullish_signal = min([to_bull_bbr_time_stamp, to_bull_engulf_time_stamp, to_ilr_time_stamp])
        if first_bullish_signal > NULL_Timestamp:
            s["To_Bull_0"] = s["Entry_Price"] - entry_window.loc[first_bullish_signal]["Close"]
        else:
            s["To_Bull_0"] = 0

        # how man times was resistance tested before signal
        r_test = before_signal_window.query("S_R == 1")["S_R"].count() 
        # how man times was resistance broken before signal
        r_break = before_signal_window.query("S_R == 2")["S_R"].count()
        # how man times was support tested before signal
        s_test = before_signal_window.query("S_R == 3")["S_R"].count()
        # how man times was support broken before signal
        s_break = before_signal_window.query("S_R == 4")["S_R"].count()

        h1_range = window[:START].iloc[-4:]["High"].max() - window[:START].iloc[-4:]["Low"].min()


        # append other useful df metrics to new df
        s["Range"] = row["Range"]
        s["ATR"] = row["ATR"]
        s["Close_%High"] = row["Close_%High"]
        s["SMA_Trend"] = row["SMA_Trend"]
        s["SMA16_Slope"] = row["SMA16_Slope"]
        s["SMA32_Slope"] = row["SMA32_Slope"]
        s["RSI"] = row["RSI"]
        s["IHR"] = row["IHR"]
        s["Dark_Cloud"] = row["Dark_Cloud"]
        s["Shooting_Star"] = row["Shooting_Star"]
        s["Bear_Engulf"] = row["Bear_Engulf"]
        s["Iday_Range"] = row["Iday_Range"]
        s["ADR"] = row["ADR"]
        s["Close_%DHigh"] = row["Close_%DHigh"]
        s["SMA16"] = row["SMA16"]
        s["SMA32"] = row["SMA32"]
        s["S_R"] = row["S_R"]
        s["YRA_Diff"] = row["YRA_Diff"]
        s["YRA_STD"] = row["YRA_STD"]
        s["R_Tested"] = r_test
        s["R_Broken"] = r_break
        s["S_Tested"] = s_test
        s["S_Broken"] = s_break
        s["H1_Range"] = h1_range
        

        signal.append(s)
    sdf = pd.DataFrame(signal)
    return sdf


### Get Bear_BBR Metrics

In [551]:
signal_idx_list = bbu_reversal.query("Bear_BBR == True")["Idx"].to_list()
bearish_bbr_df = bearish_bbr_metrics(bbu_reversal, signal_idx_list)
bearish_bbr_df.tail(10)

,Date,Entry_Price,D_Max_Up,D_Max_Down,D_Gain,Stop_Pips,Max_Up,Max_Down,To_SMA16,To_SMA32,To_BBL_16_2,To_Sig_Low,To_Bull_BBR,To_Bull_Engulf,To_ILR,To_Bull_0,Range,ATR,Close_%High,SMA_Trend,SMA16_Slope,SMA32_Slope,RSI,IHR,Dark_Cloud,Shooting_Star,Bear_Engulf,Iday_Range,ADR,Close_%DHigh,SMA16,SMA32,S_R,YRA_Diff,YRA_STD,R_Tested,R_Broken,S_Tested,S_Broken,H1_Range,To_S_0,To_S_1,To_S_2
349,2026-03-04 20:00:00-05:00,1.338030,-0.000680,0.008280,0.002145,-0.000415,-0.000635,0.000015,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000815,0.000542,0.509202,2,10.042024,15.320750,57.892203,NaN,NaN,NaN,NaN,0.002045,0.010132,0.239609,1.337425,1.336817,NaN,-0.000132,0.003961,0,0,0,0,0.001470,NaN,NaN,NaN
350,2026-03-04 20:30:00-05:00,1.336975,-0.000705,0.007225,0.001090,0.000000,-0.000260,0.007225,-0.000480,-0.000002,0.000881,0.00050,0.004685,0.000495,0.000480,0.000480,0.001820,0.000709,0.942308,2,5.651495,28.118933,48.294997,NaN,NaN,NaN,NaN,0.002235,0.010132,0.776286,1.337455,1.336977,NaN,-0.000132,0.003961,0,0,0,0,0.001840,0.000500,0.00638,NaN
351,2026-03-05 04:15:00-05:00,1.336125,-0.001260,0.006375,0.000240,0.000000,-0.000595,0.006375,0.001925,0.002301,0.002326,0.00553,0.003835,-0.000355,0.005145,-0.000355,0.002295,0.001823,0.577342,1,52.342481,-19.157014,56.943014,NaN,NaN,NaN,NaN,0.008115,0.010132,0.318546,1.333288,1.334108,NaN,-0.000132,0.003961,1,1,0,2,0.006155,0.005530,NaN,NaN
352,2026-03-05 20:15:00-05:00,1.335575,-0.006055,0.004420,-0.004505,-0.001080,-0.001145,0.000000,-0.000215,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.001195,0.000619,0.903766,2,5.828753,40.948449,53.285765,NaN,NaN,NaN,NaN,0.002125,0.010079,0.541176,1.335790,1.334766,NaN,-0.001119,0.003966,0,0,0,0,0.001265,NaN,NaN,NaN
353,2026-03-06 08:45:00-05:00,1.333850,-0.007780,0.001320,-0.006230,-0.003810,-0.004130,0.001320,0.000277,-0.000815,0.000000,-0.00075,0.000000,0.000000,0.000000,0.000000,0.004115,0.001975,0.925881,1,52.073538,-4.111065,48.065741,NaN,NaN,NaN,NaN,0.007940,0.010079,0.660579,1.333573,1.334665,4.0,-0.001119,0.003966,1,2,0,2,0.006520,-0.000750,NaN,NaN
354,2026-03-06 16:15:00-05:00,1.339925,-0.001705,0.000855,-0.000155,-0.000345,-0.000155,0.000855,0.000737,0.000000,0.000000,0.00000,0.000000,-0.000155,0.000000,0.000000,0.000360,0.001030,0.958333,2,19.978314,46.297646,58.904509,NaN,NaN,NaN,NaN,0.009825,0.010079,0.107379,1.339091,1.337920,NaN,-0.001119,0.003966,5,6,0,6,0.002405,NaN,NaN,NaN
355,2026-03-09 02:00:00-04:00,1.332730,-0.011985,0.002475,-0.010660,-0.001345,-0.001735,0.002110,0.001554,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.001590,0.001371,0.845912,2,59.302734,0.835504,55.186083,NaN,True,NaN,NaN,0.006865,0.009884,0.356154,1.330734,1.330619,NaN,0.000591,0.003713,0,0,0,0,0.003940,NaN,NaN,NaN
356,2026-03-09 09:30:00-04:00,1.337575,-0.007140,0.002990,-0.005815,-0.000695,-0.002150,0.002990,0.002020,0.000000,0.000000,0.00000,0.000000,0.000460,0.000000,0.000000,0.001580,0.001383,0.439873,2,56.383248,46.819677,66.050386,NaN,NaN,NaN,NaN,0.009960,0.009884,0.069779,1.335188,1.333955,NaN,0.000591,0.003713,1,2,1,1,0.003535,0.009265,NaN,NaN
357,2026-03-09 11:15:00-04:00,1.339030,-0.005685,0.000830,-0.004360,-0.000750,-0.001385,0.000680,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.001240,0.001635,0.604839,2,60.917219,56.519957,65.245502,NaN,NaN,NaN,NaN,0.011760,0.009884,0.088435,1.336594,1.335089,NaN,0.000591,0.003713,2,3,2,2,0.004560,NaN,NaN,NaN
358,2026-03-10 02:45:00-04:00,1.345030,-0.003310,0.003805,0.003225,-0.000745,-0.002365,0.000320,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.001135,0.000777,0.656388,1,49.774958,26.133723,63.597555,NaN,True,NaN,NaN,0.004415,0.010196,0.168743,1.343000,1.343020,3.0,0.006209,0.003829,0,2,2,1,0.002185,NaN,NaN,NaN


## Signal Stats

### Initial Signal Stats

In [465]:
# get the total number of bear_bbr signals
total_signals_count = bearish_bbr_df[bearish_bbr_df.columns[0]].count()

day_closed_lower = bearish_bbr_df.query("D_Gain > 0")
day_closed_higher = bearish_bbr_df.query("D_Gain <= 0")
day_closed_lower_count = day_closed_lower["D_Gain"].count()
day_closed_higher_count = day_closed_higher["D_Gain"].count()
day_closed_lower_pct = round(day_closed_lower_count / total_signals_count * 100,2)
day_closed_lower_pips = bearish_bbr_df["D_Gain"].sum().round(6)

# filter the dataframe for occurences where after the signal
# the price moved lower more than it moved higher
d_max_down_greater = bearish_bbr_df.query("D_Max_Up.abs() < D_Max_Down")
# get the count of occurences where price dropped more then it rose after signal
d_max_down_greater_count = d_max_down_greater[d_max_down_greater.columns[0]].count()

# percentage of where price fell further than the max it went against the entry price
d_max_down_greater_pct = round(d_max_down_greater_count / total_signals_count * 100, 2)

max_down_win = bearish_bbr_df.query("Max_Down > Max_Up.abs()")["Date"].count()
max_down_loss = bearish_bbr_df.query("Max_Down <= Max_Up.abs()")["Date"].count()
total_max_down_trades = max_down_win + max_down_loss
max_down_win_rate = round((max_down_win / total_max_down_trades) * 100 ,2)

max_down_pips = bearish_bbr_df["Max_Down"].sum().round(6)
max_up_pips = bearish_bbr_df["Max_Up"].sum().round(6)
total_max_down_gain = max_down_pips + max_up_pips

average_max_up = bearish_bbr_df["Max_Up"].mean().round(6)
average_max_down = bearish_bbr_df["Max_Down"].mean().round(6)


print(
    f"""
    ========= Initial Bear_BBR Statistics =============

    Total Bear_BBR Signals: {total_signals_count}

    ** Where the move down from the signal resulted **
    ** in a positive gain at the close of the day **
    Day Closed Lower (Count): {day_closed_lower_count}
    Day Closed Higher (Count): {day_closed_higher_count}
    Day Closed Lower (%): {day_closed_lower_pct}
    Day Closed Lower (Pips): {day_closed_lower_pips}

    ** Where after the signal the max move down on the day **
    ** was greater than the max move up **
    D_Max_Down > D_Max_Up (Count): {d_max_down_greater_count}
    D_Max_Down > D_Max_Up (%): {d_max_down_greater_pct}

    ** Where the max move down after the signal **
    ** was greater than the max move up (exit if price closes above signal high) **
    Max_Down > Max_Up (Count): {max_down_win}
    Max_Down <= Max_Up (Count): {max_down_loss}
    Total Trades (Max_Down): {total_max_down_trades}
    Max_Down > Max_Up (Win Rate): {max_down_win_rate}

    Max_Down Pips: {max_down_pips}
    Max_Up Pips: {max_up_pips}
    Total Max_Down Gain: {total_max_down_gain}
    
    Average Max_Up Pips: {average_max_up}
    Average Max_Down Pips: {average_max_down}
 
    ==============================================================================
    """
)


    ========= Initial Bear_BBR Statistics =============

    Total Bear_BBR Signals: 359

    ** Where the move down from the signal resulted **
    ** in a positive gain at the close of the day **
    Day Closed Lower (Count): 172
    Day Closed Higher (Count): 187
    Day Closed Lower (%): 47.91
    Day Closed Lower (Pips): -0.028125

    ** Where after the signal the max move down on the day **
    ** was greater than the max move up **
    D_Max_Down > D_Max_Up (Count): 171
    D_Max_Down > D_Max_Up (%): 47.63

    ** Where the max move down after the signal **
    ** was greater than the max move up (exit if price closes above signal high) **
    Max_Down > Max_Up (Count): 166
    Max_Down <= Max_Up (Count): 190
    Total Trades (Max_Down): 356
    Max_Down > Max_Up (Win Rate): 46.63

    Max_Down Pips: 0.728515
    Max_Up Pips: -0.36732
    Total Max_Down Gain: 0.36119500000000004
    
    Average Max_Up Pips: -0.001032
    Average Max_Down Pips: 0.002046
 
    


### Summary

- From the initial numbers, we can see that the Bear_BBR signal occurred `359` times.
- After the signal occuring, the *max price move down* was greater than the *max price move up* on the day, `47.63%` of the time.
- When the signal occurred, `46.63%` of the time the price dropped more than it gained after the signal, before closing above the signal high.
- The total gains where the price fell more than it rose after the signal (before closing above the signal high) was `3611` pips.
- The average number of pips where price moved against the signal was `10.32` pips
- The average number of pips gained before price closed above the signal high was `20.46` pips.

#### What does this tell us?

- We can see just under 50% of the time, the price moves lower after the signal, *before or without closing above the signal high*.
- Additionally, the gains from when the price moves in favour of the signal is greater than when the price moves against the signal. In short the signal generates more gains than losses overall.
- On average, the price moves down nearly double the amount it moves against the signal.

#### What information is missing?

- Although the signal gains more pips overall, this doesn't tell us the best place to take profit to capture the max move down.
- It is not clear what factors make a move down more likely, or when the signal is more likely to fail.


## Investigating factors that increase signal accuracy

### Holding a short position from signal entry to close of the day



In [468]:
def daily_pip_gain(df: Series, pct_adr: int):
    d_pip_gain = df["D_Gain"] if abs(df["D_Max_Up"]) < (df["ADR"] * pct_adr) else (df["ADR"] * -pct_adr)
    return d_pip_gain

def condition_stats(df: DataFrame, pct_adr: int):

    df["Pip_Gain"] = df.apply(daily_pip_gain, axis=1, args=[pct_adr])
    df["ADR%"] = df.apply(lambda x: x["ADR"] * pct_adr, axis=1)
    
    success_series = df.query(f"D_Max_Up.abs() < (ADR * {pct_adr})")["Pip_Gain"]
    fail_series = df.query(f"D_Max_Up.abs() >= (ADR * {pct_adr})")["Pip_Gain"]

    success_total = success_series.count()
    fail_total = fail_series.count()
    total_trades = success_total + fail_total
    
    win_rate = round(success_total / total_trades * 100,2)
    condition_total_pips = round(success_series.sum() + fail_series.sum(),6)

    return {
        "Pct_ADR": pct_adr,
        "Pct_ADR_Avg": df["ADR%"].mean(),
        "ADR_Avg": df["ADR"].mean(),
        "Total Trades": total_trades,
        "Win": success_total,
        "Loss": fail_total,
        "Win Rate": win_rate,
        "Avg_win": success_series.mean(),
        "Avg_Loss": fail_series.mean(),
        "Total Pips": condition_total_pips
    }
    
# columns
cols1 = [
    "Date", "Entry_Price", "Pip_Gain", "D_Max_Up", "ADR%", "D_Gain", "ADR"]
#df
d_gain_df = bearish_bbr_df


### Using a stop loss set to a percentage of the ADR

Since we already know from the intial stats that the price will close lower on the day 48% of the time after the signal, we should try to limit the losses for the 52% of the times where it fails.

A `stop loss` is used to limit losses, and can be static (e.g. 10 pips) or dynamic (e.g. set as multiple of the `ATR`, or a trailing stop ... etc.).
The ATR takes into account the volatility of the last *x* candles, so is usually a good choice. 

For this particular scenario however, when the `Bear_BBR` signal occurs we are looking to hold a short position until the close of the day. The ATR is not the best approach here, because ATR will look at the average candle range within increments of the timeframe period (e.g. 15min) over *x* period of candles. 

To address this I will use a percentage of the the average daily range (**ADR**) as a stop loss. The ADR calculates the average daily price movement over last 30 days of trading, which will tell me the average number of pips the price moves within a day over that period.

To find a reasonable stop loss range, I'll iterate through different ADR percentages from 1 to 100.

In [469]:
d_gain_experimental_sl = [condition_stats(d_gain_df, x * 0.01) for x in range(1,101)]
d_gain_ex_sl_res_df = pd.DataFrame(d_gain_experimental_sl)
d_gain_ex_sl_res_df

,Pct_ADR,Pct_ADR_Avg,ADR_Avg,Total Trades,Win,Loss,Win Rate,Avg_win,Avg_Loss,Total Pips
0,0.01,0.000091,0.00908,319,10,309,3.13,0.003507,-0.000091,0.007012
1,0.02,0.000182,0.00908,319,12,307,3.76,0.003565,-0.000182,-0.012991
2,0.03,0.000272,0.00908,319,23,296,7.21,0.003789,-0.000272,0.006551
3,0.04,0.000363,0.00908,319,26,293,8.15,0.003751,-0.000363,-0.008811
4,0.05,0.000454,0.00908,319,31,288,9.72,0.003717,-0.000454,-0.015567
5,0.06,0.000545,0.00908,319,39,280,12.23,0.003917,-0.000545,0.000092
6,0.07,0.000636,0.00908,319,47,272,14.73,0.003584,-0.000635,-0.004154
7,0.08,0.000726,0.00908,319,51,268,15.99,0.003519,-0.000725,-0.014929
8,0.09,0.000817,0.00908,319,54,265,16.93,0.003467,-0.000816,-0.029098
9,0.10,0.000908,0.00908,319,59,260,18.50,0.003707,-0.000907,-0.016967


- This shows the stop_loss is directly related to the win rate.
- It also shows that even with a higher win rate, it does not mean a higher `total pips` value. This makes sense because the larger the stop loss, the larger each loss will be.
- The ideal stop loss will be as low as possible and yield a reasonably high `total pips`.
- The maximum total pips returned was 0.073820 which was a result of 28% ADR stop loss, however we can't take this at face value because it could be an outlier. We also need to consider that when a trade is placed, a percentage of equity will be risked (e.g. 1%), so a lower ADR percentage will allow for a larger trade size to be used, which maximises the monetary return.
- We should look at the top ten `total pips` values and see what range of ADR percentages provided the best return.

In [470]:
top_ten_returns_by_adr_pct = d_gain_ex_sl_res_df["Total Pips"].nlargest(10)
top_ten_returns_by_adr_pct

27    0.073820
15    0.068408
47    0.066729
31    0.065609
32    0.064584
48    0.061723
44    0.060313
28    0.059654
26    0.059358
16    0.059083
Name: Total Pips, dtype: float64

The output shows the top ten `total pips` values by their index. Since the percent of ADR value is the `index+1`, we can see what contiguous index values provide the best return. I prefer contiguous values to get a better approximation of ADR percent ranges that work well, which should help to avoid outliers that may have been a fluke. 

From the output there's 3 contiguous index values between 26-28 which map to 27-29% ADR, another 3 at 45% and 48-49%. Since the 40's are not contiguous values I'll discount them. That leaves 2 from the 30's (32-33%), and 2 from the 10's (16-17.

Overall, the larger contiguous spread of the highest `total pips` is between 32 and 33%. The win rate is 55.49%, 56.11% however, the average win (0.002702, 0.002701) is less than the average loss (0.002906, 0.002992). This means you would have to rely on a achieving a similar win rate to be profitable. If the win rate drops by 6% you will start loosing money. The average loss for 32-33% ADR is also more than the average for 27-29% ADR, which has larger average wins than losses.

When you look at 16-17% ADR in comparison, the win rate is (32.92, 34.80) but the average win (0.003602, 0.003417) is double the average loss (0.001448, 0.001539). The 16% ADR value also yields the second highest `total pips` return out of all ADR percentage values. So 16% ADR looks good becauses it keeps the pip loss small, and still has a reasonable win rate which we can try to improve.

The reason for choosing the lower ADR percentage value is that is allows us to take a larger trade size while risking the same equity percentage. A win with a larger trade value will generate a larger monetary return for the total pips. To see this in action, I'll simulate an account size of $100,000 and a trade risk value of 1%. 

In [471]:
risk_ex_df = d_gain_ex_sl_res_df
risk_ex_df["Account_Size"] = risk_ex_df.apply(lambda x: 100000, axis=1)
risk_ex_df["Risk"] = risk_ex_df.apply(lambda x: 0.01, axis=1)
risk_ex_df["Trade_Value"] = risk_ex_df.apply(lambda x: round((x["Account_Size"] * x["Risk"])/x["Pct_ADR_Avg"],2), axis=1)
risk_ex_df["Return"] = risk_ex_df.apply(lambda x: round(x["Trade_Value"] * x["Total Pips"],2), axis=1)

risk_ex_df.filter(risk_ex_df["Return"].nlargest(10).index, axis=0)

,Pct_ADR,Pct_ADR_Avg,ADR_Avg,Total Trades,Win,Loss,Win Rate,Avg_win,Avg_Loss,Total Pips,Account_Size,Risk,Trade_Value,Return
0,0.01,0.000091,0.00908,319,10,309,3.13,0.003507,-0.000091,0.007012,100000,0.01,11013193.26,77224.51
15,0.16,0.001453,0.00908,319,105,214,32.92,0.003602,-0.001448,0.068408,100000,0.01,688324.58,47086.91
16,0.17,0.001544,0.00908,319,111,208,34.80,0.003417,-0.001539,0.059083,100000,0.01,647834.90,38276.03
17,0.18,0.001634,0.00908,319,116,203,36.36,0.003313,-0.001628,0.053708,100000,0.01,611844.07,32860.92
27,0.28,0.002542,0.00908,319,163,156,51.10,0.002886,-0.002543,0.073820,100000,0.01,393328.33,29035.50
14,0.15,0.001362,0.00908,319,91,228,28.53,0.003812,-0.001360,0.036870,100000,0.01,734212.88,27070.43
19,0.20,0.001816,0.00908,319,123,196,38.56,0.003272,-0.001811,0.047553,100000,0.01,550659.66,26185.52
21,0.22,0.001998,0.00908,319,136,183,42.63,0.003048,-0.001989,0.050539,100000,0.01,500599.69,25299.81
18,0.19,0.001725,0.00908,319,119,200,37.30,0.003255,-0.001719,0.043531,100000,0.01,579641.75,25232.39
26,0.27,0.002452,0.00908,319,156,163,48.90,0.002945,-0.002454,0.059358,100000,0.01,407896.05,24211.89


The results show an ADR percentage of 1% yields the highest return but has a 3% win rate. This is clearly an anomaly because 2 to 3% ADR would also yield high returns if this was a good area for a stop loss. 

The better indicator is the range between 16-18%, which is a countiguous range of values with the highest monetary return out of the distribution. Based on this I'll view 16% ADR as an ideal stop loss and look for ways to improve the win rate.

### Testing if the slope angle increases win rate

Considering the Bear_BBR is a short signal, we want to avoid cases were there is a strong uptrend. There's probably many ways to define a strong uptrend, but I think looking at the angle of the simple moving average (SMA) slopes is a way to define it. Generally, a strong trend or up-move will have SMA's pointing up. To quantify "pointing up" we'll say if the angle is greater than *x* the trend is too bullish, so skip the signal.

I think setting the angle limit to 45 degrees makes sense. Let's see what this shows:

In [472]:
# conditions
slope_lt_45_df = bearish_bbr_df.query("SMA32_Slope < 45").copy()
slope_lt_45_res = condition_stats(slope_lt_45_df, 0.28)
pd.DataFrame([slope_lt_45_res])

,Pct_ADR,Pct_ADR_Avg,ADR_Avg,Total Trades,Win,Loss,Win Rate,Avg_win,Avg_Loss,Total Pips
0,0.28,0.002533,0.009048,288,150,138,52.08,0.003041,-0.002537,0.105919


This shows a slight reduction in the total number of trades (288), which we'd expect with this filter. We also see 2% increase in the win rate, and an increase of 230 pips, making the `total pips` gained 909. It definitely looks better, but we can also test for a range of angles betweem 90 and 0, to see of there is anything better.

In [473]:
slope_angle_ex = [condition_stats(bearish_bbr_df.query(f"SMA32_Slope < {x}").copy(), 0.28) for x in range(0,91)]
slope_angle_ex_df = pd.DataFrame(slope_angle_ex)
slope_angle_ex_df

,Pct_ADR,Pct_ADR_Avg,ADR_Avg,Total Trades,Win,Loss,Win Rate,Avg_win,Avg_Loss,Total Pips
0,0.28,0.002529,0.009031,86,37,49,43.02,0.002283,-0.002548,-0.040353
1,0.28,0.002527,0.009024,91,40,51,43.96,0.002301,-0.002554,-0.038212
2,0.28,0.002520,0.009001,94,42,52,44.68,0.002434,-0.002548,-0.030244
3,0.28,0.002517,0.008988,100,46,54,46.00,0.002476,-0.002533,-0.022896
4,0.28,0.002514,0.008977,102,48,54,47.06,0.002521,-0.002533,-0.015756
5,0.28,0.002514,0.008977,104,49,55,47.12,0.002546,-0.002528,-0.014264
6,0.28,0.002517,0.008988,105,50,55,47.62,0.002615,-0.002528,-0.008289
7,0.28,0.002510,0.008965,110,55,55,50.00,0.002460,-0.002528,-0.003694
8,0.28,0.002515,0.008982,111,56,55,50.45,0.002435,-0.002528,-0.002654
9,0.28,0.002517,0.008988,122,63,59,51.64,0.002811,-0.002515,0.028691


We can view the angles with the largest returns:

In [474]:
slope_angle_ex_df["Account_Size"] = slope_angle_ex_df.apply(lambda x: 100000, axis=1)
slope_angle_ex_df["Risk"] = slope_angle_ex_df.apply(lambda x: 0.01, axis=1)
slope_angle_ex_df["Trade_Value"] = slope_angle_ex_df.apply(lambda x: round((x["Account_Size"] * x["Risk"])/x["Pct_ADR_Avg"],2), axis=1)
slope_angle_ex_df["Return"] = slope_angle_ex_df.apply(lambda x: round(x["Trade_Value"] * x["Total Pips"],2), axis=1)
slope_angle_ex_df.filter(slope_angle_ex_df["Total Pips"].nlargest(10).index, axis=0)

,Pct_ADR,Pct_ADR_Avg,ADR_Avg,Total Trades,Win,Loss,Win Rate,Avg_win,Avg_Loss,Total Pips,Account_Size,Risk,Trade_Value,Return
33,0.28,0.002521,0.009005,257,135,122,52.53,0.003130,-0.002535,0.113330,100000,0.01,396592.86,44945.87
40,0.28,0.002528,0.009028,282,148,134,52.48,0.003053,-0.002528,0.113083,100000,0.01,395588.66,44734.35
41,0.28,0.002529,0.009032,283,148,135,52.30,0.003053,-0.002531,0.110261,100000,0.01,395427.73,43600.26
34,0.28,0.002522,0.009006,260,137,123,52.69,0.003073,-0.002536,0.109081,100000,0.01,396566.72,43257.89
44,0.28,0.002533,0.009048,288,150,138,52.08,0.003041,-0.002537,0.105919,100000,0.01,394736.86,41810.13
45,0.28,0.002533,0.009048,288,150,138,52.08,0.003041,-0.002537,0.105919,100000,0.01,394736.86,41810.13
46,0.28,0.002533,0.009048,288,150,138,52.08,0.003041,-0.002537,0.105919,100000,0.01,394736.86,41810.13
43,0.28,0.002530,0.009035,286,149,137,52.10,0.003035,-0.002532,0.105377,100000,0.01,395274.99,41652.89
42,0.28,0.002529,0.009033,285,148,137,51.93,0.003053,-0.002532,0.105077,100000,0.01,395359.42,41543.18
26,0.28,0.002517,0.008990,213,114,99,53.52,0.003093,-0.002516,0.103511,100000,0.01,397278.88,41122.73


This shows angles between the range of 33 and 42 have returns above 900 pips, however the best fit is **33** degrees. From this sample, if we only took `Bear_BBR` trades with an SMA32 slope under 33 degrees the `total pips` would be 990 with a monetary return of $68703 (excluding fees and commissions). 

In [477]:
#p = pd.Series([x * 0.01 for x in range(1,100)])
#test_range_ex = [condition_stats(bearish_bbr_df.query(f"SMA32_Slope < 45 and Iday_Range < ADR * {x}").copy(), 0.16) for x in p] 
#test_range_ex_df = pd.DataFrame(test_range_ex)
#test_range_ex_df.filter(test_range_ex_df["Total Pips"].nlargest(10).index, axis=0)
sr_ex = bearish_bbr_df
sr_ex_list = [condition_stats(sr_ex.query("SMA32_Slope < 33 ").copy(), 0.28)]
sr_ex_df = pd.DataFrame(sr_ex_list)
sr_ex_df

,Pct_ADR,Pct_ADR_Avg,ADR_Avg,Total Trades,Win,Loss,Win Rate,Avg_win,Avg_Loss,Total Pips
0,0.28,0.002521,0.009005,257,135,122,52.53,0.00313,-0.002535,0.11333
